rust/src/lib.rs

```diff
@@ -26,7 +26,11 @@
 #![cfg_attr(feature = "bench", feature(test))]
 #![cfg_attr(feature = "staticlib", feature(lang_items))]
 
+#[cfg(features = "noni")]
+extern crate aes_soft as aesni;
+#[cfg(not(features = "noni"))]
 extern crate aesni;
```

Comment:
Per the above comment about gating on the crates themselves, this could be something like:

```rust
#[cfg(features = "aes-soft")]
extern crate aes_soft as aes;
#[cfg(features = "aesni"))]
extern crate aesni as aes;
```

(with corresponding changes to all `use aesni` directives)

Attempting to use both would cause a (perhaps counterintuitive) compile error, but hopefully we can document the usage patterns in ways people don't attempt that.

Reply:
You can add the following line before those two:
```Rust
#[cfg(all(features = "aesni", features = "aes-soft")))]
compile_errror!("aesni and aes-soft features are mutually exclusive");
```

Reply:
I just rebased this PR, I'll add the changes you outlined in your review next :+1: